### PR TITLE
`Exam mode`: Allow instructors to use submission policies

### PR DESCRIPTION
**Pull request by krusche**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [ ] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [ ] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [ ] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.

#### Changes affecting Programming Exercises
- [ ] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [ ] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

#### Exam Mode Testing

Prerequisites:
- 1 Student
- 1 Exam with a Programming Exercise with submission policy enabled

1. Log in to Artemis as student
2. Participate in the exam as a student
3. Make sure that the UI of the programming exercise shows information about the submission policy status (see screenshots below)

#### Performance Review
- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
#### Exam Mode Test
- [ ] Test 1
- [ ] Test 2

### Screenshots
If the submission policy is set to 'Submission Penalty':
policy_label

If the submission policy is set to 'Lock Repository' and the student exceeds the number of allowed submissions:
lock

If the online code editor is disabled, submission policy is set to 'Lock Repository', and the student exceeds the number of allowed submissions:
no-editor-lock

### CODE REVIEW

File: src/main/webapp/app/exam/participate/exercises/programming/programming-exam-submission.component.ts [hunk 1]
```diff
@@ -114,6 +137,18 @@ export class ProgrammingExamSubmissionComponent extends ExamSubmissionComponent
...
+    /**
+     * Called when the build is completed after a new commit.
+     * Intended only for the case when the online editor is disabled (otherwise, we can update on CommitState changes)
+     */
+    onBuildComplete() {
+        if (this.exercise.allowOnlineEditor || !this.exercise.allowOfflineIde) {
+            return;
+        }
+        // this.setSubmissionCountAndLockIfNeeded(this.studentParticipation.submissionCount);
+        // FIXME: this.studentParticipation contains the old submissionCount value here -> can't update submissionCount
```
Comment (terlan98) [inline src/main/webapp/app/exam/participate/exercises/programming/programming-exam-submission.component.ts:149]: Not sure how to handle this. Currently, when the code editor is disabled the only way to update `submissionCount` is to refresh the page. 

Should I increment `submissionCount` locally and compare it to `exercise.submissionPolicy.submissionLimit` to determine if the repo should be locked? 🤔

Comment (krusche) [reply src/main/webapp/app/exam/participate/exercises/programming/programming-exam-submission.component.ts:149]: When the student pushes code, the client should receive a websocket message with the programming submission. This should contain a link to the participation with the submission count and locked=true/false

So the server is already calculating the required information and you need to extract and update the local objects with the values from the websocket message. Can you try if this is possible? (Hopefully without too many changes)

**Review by laadvo — COMMENTED** [review]
Tested on ts1, the number of submissions was correctly displayed for both submission policies. For the submission policy "locked repository", when the submission limit is reached, the submit button is disabled, however you are still able to submit by cloning the project and pushing the changes. 
image

 I was not able to test if the submission penalty is applied correctly after reaching the limit since programming exercises are currently not building.

**Review by DominikRemo — COMMENTED** [review]
Tested on ts1, same problem as @laadvo

**Comment by terlan98** [comment]
@laadvo @DominikRemo @RY997, thanks a lot for testing! Apparently, the repositories are locked when the result of a submission is received and since there's an issue with the build agents right now, repositories are not locked. 

We know this is not ideal since there can be a long delay between a submission and its result, but it will be fixed in a follow-up PR. If there are no other issues, please approve 🙂

**Review by laadvo — APPROVED** [review]
Other than the problem with repository locking everthing worked :+1:

**Review by RY997 — APPROVED** [review]
Worked as expected

**Review by DominikRemo — APPROVED** [review]
Apart from the mentioned issue, everything worked well

**Review by laurenzfb — APPROVED** [review]
Looked through the code, LGTM

**Review by MaximilianAnzinger — APPROVED** [review]
Overall code looks good to me. Left some suggestions/questions.

**Review by MichaelOwenDyer — APPROVED** [review]
Changes all look good to me 👍🏽

**Review by dearjasmina — APPROVED** [review]
Code looks good

**Pull request merged**